Readme update (#37)
morpho-org/public-allocator: README.md

# Public Allocator

## Overview

[MetaMorpho](https://github.com/morpho-org/metamorpho) is a protocol for funds allocations and risk management on top of [Morpho Blue](https://github.com/morpho-org/morpho-blue). The Public Allocator is a contract that MetaMorpho curators can enabled as [allocator](https://github.com/morpho-org/metamorpho?tab=readme-ov-file#allocator) to let anybody reallocate the vault's funds to fill their liquidity needs.

> The Public Allocator's function `reallocateTo` ([source](https://github.com/morpho-org/public-allocator/blob/a43e2e29ba425a5e2af3dcb7c6774a4524abcda3/src/PublicAllocator.sol#L108), [interface](https://github.com/morpho-org/public-allocator/blob/cd82f57f2ca52c71fdb44ffb14278844beb593ec/src/interfaces/IPublicAllocator.sol#L62)) reallocates some liquidity from multiple markets to one market. It takes as input `withdrawals`, a list of `(MarketParams, uint128)` pairs, which will be the markets and amounts that will be reallocated, and `supplyMarketParams`, a `MarketParams`, which is the market in which to supply the reallocated funds. Note that the `id` of the markets in `withdrawals` must be sorted, and that `supplyMarketParams` cannot be a market of `withdrawals`.

Reviewer: Both of the links could be taken from the same commit. Both are a little off of a few lines, they do not redirect to the exact line that we want

Author: weird, not sure how that could happen

Author: ```suggestion
The Public Allocator's function `reallocateTo` ([source](https://github.com/morpho-org/public-allocator/blob/7271fbd60881ff32a466a588f99344c6bf72629a/src/PublicAllocator.sol#L108), [interface](https://github.com/morpho-org/public-allocator/blob/7271fbd60881ff32a466a588f99344c6bf72629a/src/interfaces/IPublicAllocator.sol#L62)) reallocates some liquidity from multiple markets to one market. It takes as input `withdrawals`, a list of `(MarketParams, uint128)` pairs, which will be the markets and amounts that will be reallocated, and `supplyMarketParams`, a `MarketParams`, which is the market in which to supply the reallocated funds. Note that the `id` of the markets in `withdrawals` must be sorted, and that `supplyMarketParams` cannot be a market of `withdrawals`.
```

> The Public Allocator's function `reallocateTo` ([source](https://github.com/morpho-org/public-allocator/blob/a43e2e29ba425a5e2af3dcb7c6774a4524abcda3/src/PublicAllocator.sol#L108), [interface](https://github.com/morpho-org/public-allocator/blob/cd82f57f2ca52c71fdb44ffb14278844beb593ec/src/interfaces/IPublicAllocator.sol#L62)) reallocates some liquidity from multiple markets to one market. It takes as input `withdrawals`, a list of `(MarketParams, uint128)` pairs, which will be the markets and amounts that will be reallocated, and `supplyMarketParams`, a `MarketParams`, which is the market in which to supply the reallocated funds. Note that the `id` of the markets in `withdrawals` must be sorted, and that `supplyMarketParams` cannot be a market of `withdrawals`.

Reviewer: Not sure if we should go into the details of the interface. Better if we can avoid to say that `withdrawals` is a list of pairs, it can be confusing as `Withdrawal` is a struct

The Public Allocator provides the possibility to the vault owner or a per vault settable admin to constrain the public reallocation:
- **Max flows**: Each market has a max inflow (`maxIn`) and max outflow (`maxOut`), that can be set by the vault owner or the Public Allocator vault admin. The markets from which funds are withdrawn through `reallocateTo` increase their `maxIn` and decrease their `maxOut`, and the market in which funds are deposited decrease their `maxOut` and increase their `maxIn`.

> - **Fee**: If set, user must pay a fee in ETH to be able to call `reallocateTo`. The vault owner or the Public Allocator vault admin can set this fee, as welle as taking the accumulated fees out by calling `transferFee`.

Reviewer: ```suggestion
- **Fee**: If set, user must pay a fee in ETH to be able to call `reallocateTo`. The vault owner or the Public Allocator vault admin can set this fee, as well as taking the accumulated fees out by calling `transferFee`.
```

## Testing

To run tests: `forge test`. Note that running the tests this way will use this repository's compilation settings for dependencies (Morpho & MetaMorpho), potentially different from the settings used for deployment of those dependencies.

## Audits

All audits are stored in the [audits](./audits/)' folder.

> ## Licence
> 
> All files are licenced under `GPL-2.0-or-later`, see [`LICENCE`](./LICENCE).

Reviewer: ```suggestion
## License

All files are licenced under `GPL-2.0-or-later`, see [`LICENSE`](./LICENSE).
```
that's what we chose for morpho-blue and metamorpho

> All files are licenced under `GPL-2.0-or-later`, see [`LICENCE`](./LICENCE).

Reviewer: That makes me think that the `LICENSE` file is missing

# Public Allocator

## Overview

[MetaMorpho](https://github.com/morpho-org/metamorpho) is a protocol for funds allocations and risk management on top of [Morpho Blue](https://github.com/morpho-org/morpho-blue).

> The Public Allocator is a contract that MetaMorpho curators can enable as [allocator](https://github.com/morpho-org/metamorpho?tab=readme-ov-file#allocator) to let anybody reallocate the vault's funds to fill their liquidity needs.

Reviewer: ```suggestion
The Public Allocator is a contract that MetaMorpho owner can enable as [allocator](https://github.com/morpho-org/metamorpho?tab=readme-ov-file#allocator) to let anybody reallocate the vault's funds to fill their liquidity needs.
```
otherwise it's too misleading with the curator role

Reviewer: later we call it "vault owner", maybe we can harmonize

Author: ```suggestion
The Public Allocator is a contract that MetaMorpho vault owners can enable as [allocator](https://github.com/morpho-org/metamorpho?tab=readme-ov-file#allocator) to let anybody reallocate the vault's funds to fill their liquidity needs.
```

Author: ```suggestion
The Public Allocator is a contract that MetaMorpho vault owners can enable as [allocator](https://github.com/morpho-org/metamorpho?tab=readme-ov-file#allocator) to let anybody reallocate the vault's funds to fill their liquidity needs.
```

The Public Allocator's function `reallocateTo` ([source](https://github.com/morpho-org/public-allocator/blob/7271fbd60881ff32a466a588f99344c6bf72629a/src/PublicAllocator.sol#L108), [interface](https://github.com/morpho-org/public-allocator/blob/7271fbd60881ff32a466a588f99344c6bf72629a/src/interfaces/IPublicAllocator.sol#L62)) reallocates some liquidity from multiple markets to one market.
It takes as input `withdrawals`, a list of pairs, which will be the markets and amounts that will be reallocated, and `supplyMarketParams`, a `MarketParams`, which is the market in which to supply the reallocated funds.

> Note that the `id` of the markets in `withdrawals` must be sorted, and that `supplyMarketParams` cannot be a market of `withdrawals`.

Reviewer 2: ```suggestion
Note that the `id` of the markets in `withdrawals` must be sorted in ascending alphabetical order, and that `supplyMarketParams` cannot be a market of `withdrawals`.
```

Reviewer: I think it's confusing to say `alphabetical` here, because it's really the usual order on hex number. For example `0xa0` is greater than `0xb` (meaning `0x0b`) but you would find it first in a dictionary

Reviewer 2: Right, just to make sure our offchain code is not completely off: for same-length hex strings, the expected order does correspond to the alphabetical order right?

Reviewer: Yes, because the alphabetical order is the same as the (usual) numerical order for numbers of the same length

Author: But what's before between 0x9 and 0xa in "alphabetical" order ? Better not to put it I think
